Add option to download images

64bitpandas opened the pull request:
I'm using this tool to mirror some of my Substack posts to my website, and as part of that process I'd really like to host my own images instead of having them link to the Substack CDN!

In case this will help someone else, here's a PR 🙂 

Here's a list of some tweaks I made to get that to happen:
- Add an `--images` flag that will download images for all posts being scraped into a `substack_images/` folder
- Add an option to download a single post (by passing in a `--url` in the format `https://example.substack.com/p/postname`
- When downloading images, Substack nests them like `[![alt](/path)](/path)`. Change these to just be `![alt](/path)` so clicking on the images doesn't link to itself.
- Add some tests, to prove to myself this code works the way I expect it


As a bonus, the progress bars reflect image downloads (since they can take a while)!  As an example:
```
Scraping posts: 100%|██████████| 2/2 [00:30<00:00, 15.00s/post]
  Downloading images for test-post: 100%|██████████| 7/7 [00:14<00:00, 2.00s/image]
  Downloading images for another-post: 100%|██████████| 4/4 [00:08<00:00, 2.00s/image]
```

milahu commented:
commit based on this PR: https://github.com/milahu/substack2markdown/commit/5811bb5de77a0302e619936afaa7fbe67d45b307

milahu commented on substack_scraper.py:
```diff
@@ -17,23 +22,134 @@
+    if "substackcdn.com" in url:
+        # Get the actual image URL after the CDN parameters
+        original_url = unquote(url.split("https://")[1])
+        filename = original_url.split("/")[-1]
```
`filename` can be wrong, because `substackcdn.com/image/fetch` can

- change the image size (crop) (ex: 1536x1024 &rarr; 1456x971)
- change the image format (ex: png &rarr; webp)

also, when a post uses the same original image in different sizes
then all image versions use the same filename (filename collisions)

example

```
cd  $(mktemp -d)

$ wget 'https://substackcdn.com/image/fetch/$s_!UkZH!,w_1456,c_limit,f_webp,q_auto:good,fl_progressive:steep/https%3A%2F%2Fsubstack-post-media.s3.amazonaws.com%2Fpublic%2Fimages%2Fd2c2394c-643c-4533-8d74-734252ae02ac_1536x1024.png'

$ mv * d2c2394c-643c-4533-8d74-734252ae02ac_1536x1024.png.webp

$ python
>>> import urllib.parse
>>> urllib.parse.unquote('https://substackcdn.com/image/fetch/$s_!UkZH!,w_1456,c_limit,f_webp,q_auto:good,fl_progressive:steep/https%3A%2F%2Fsubstack-post-media.s3.amazonaws.com%2Fpublic%2Fimages%2Fd2c2394c-643c-4533-8d74-734252ae02ac_1536x1024.png')
'https://substackcdn.com/image/fetch/$s_!UkZH!,w_1456,c_limit,f_webp,q_auto:good,fl_progressive:steep/https://substack-post-media.s3.amazonaws.com/public/images/d2c2394c-643c-4533-8d74-734252ae02ac_1536x1024.png'

$ wget 'https://substack-post-media.s3.amazonaws.com/public/images/d2c2394c-643c-4533-8d74-734252ae02ac_1536x1024.png'

$ identify *
d2c2394c-643c-4533-8d74-734252ae02ac_1536x1024.png PNG 1536x1024 1536x1024+0+0 8-bit sRGB 2.56234MiB 0.000u 0:00.000
d2c2394c-643c-4533-8d74-734252ae02ac_1536x1024.png.webp WEBP 1456x971 1456x971+0+0 8-bit sRGB 207452B 0.000u 0:00.000
```

simple fix: always download the original image

pro: simple
con: the original images can be huge! in my case 10x larger

```py
def resolve_image_url(url: str) -> str:
    """Get the original image URL."""
    # https://substackcdn.com/image/fetch/xxx/https%3A%2F%2Fsubstack-post-media.s3.amazonaws.com%2Fpublic%2Fimages%2Fxxx
    if url.startswith("https://substackcdn.com/image/fetch/"):
        # substackcdn.com returns a compressed version of the original image
        url = "https://" + unquote(url.split("/https%3A%2F%2F")[1])
    return url
```

```py
            url = resolve_image_url(url)
            filename = sanitize_image_filename(url)
```

timf34 commented:
Thanks for this PR @64bitpandas! The image downloading feature is useful, and I especially appreciate that you included tests — that's rare for this repo.

Unfortunately the codebase has changed significantly since this was opened (new `BrowserManager` class, refactored CLI, etc.), and this PR now has merge conflicts. There's also a small bug at the end of `main()` where `scraper.scrape_posts(args.number)` gets called twice.

I'd love to see this feature re-submitted as a fresh PR against the current codebase if you're still interested. The core idea (download images, update markdown paths, `--images` flag) is solid and could probably be done with less code now.

I'll close this in about a week unless you'd like to discuss or rebase. Thanks for the contribution!

64bitpandas committed:
feat: add image downloading and single post URL support

Agent-Id: agent-bccd3d96-55bf-4d1c-aa4b-dc85509b6986
Linked-Note-Id: b72e70ac-49ae-4300-acb1-528eb13a2043

64bitpandas committed:
test: add comprehensive test suite for substack scraper

Agent-Id: agent-558e79bc-f991-4265-8c0a-cf8665d2226e
Linked-Note-Id: ccff6987-eeca-471e-a904-1b0f42b65117

64bitpandas committed:
docs: update README, .gitignore, and requirements for image download feature

Agent-Id: agent-ec649ac2-bf40-4573-ac97-d4218ed9a2f8

64bitpandas commented:
> Thanks for this PR @64bitpandas! The image downloading feature is useful, and I especially appreciate that you included tests — that's rare for this repo.
> 
> Unfortunately the codebase has changed significantly since this was opened (new `BrowserManager` class, refactored CLI, etc.), and this PR now has merge conflicts. There's also a small bug at the end of `main()` where `scraper.scrape_posts(args.number)` gets called twice.
> 
> I'd love to see this feature re-submitted as a fresh PR against the current codebase if you're still interested. The core idea (download images, update markdown paths, `--images` flag) is solid and could probably be done with less code now.
> 
> I'll close this in about a week unless you'd like to discuss or rebase. Thanks for the contribution!

rebased + addressed @milahu 's comment above, lmk how it looks!

timf34 commented:
@64bitpandas this is excellent thank you so much! And apologies that I've been so slow with this PR - thank you very much for the swift effort.
I'm sure many users will really appreciate this!
Best,
Tim
